File: ios/Touliao/Core/Realtime/SocialReadGuard.swift
import Foundation

/// Request ordering plus identity epoch; events contain no values to merge.
@MainActor
final class SocialReadGuard {
    struct Stamp { let sequence: UInt64; let revision: UInt64 }
    private let identity: () -> UInt64
    private let revision: () -> UInt64
    private let owner: UInt64
    private var sequence: UInt64 = 0
    init(identity: @escaping () -> UInt64 = { KeychainStore.shared.snapshot().identityEpoch },
         revision: @escaping () -> UInt64 = { SocketService.shared.socialState.value }) {
        self.identity = identity; self.revision = revision; owner = identity()
    }
    func begin() -> Stamp? {
        guard identity() == owner else { return nil }
        sequence += 1
        return Stamp(sequence: sequence, revision: revision())
    }
    func current(_ stamp: Stamp) -> Bool { identity() == owner && sequence == stamp.sequence && revision() == stamp.revision }
}

File: android/app/src/main/java/com/touliao/app/core/realtime/SocialReadGuard.kt
package com.touliao.app.core.realtime

/** A view belongs to one identity epoch. Invalidations fence pending GETs. */
class SocialReadGuard(private val identity: () -> Long, private val revision: () -> Long) {
    private val owner = identity()
    private var sequence = 0L
    data class Stamp(val sequence: Long, val revision: Long)
    fun begin(): Stamp? = if (identity() != owner) null else Stamp(++sequence, revision())
    fun current(stamp: Stamp): Boolean = identity() == owner && sequence == stamp.sequence && revision() == stamp.revision
}
